From 9d50fdac4da5eb925eea4ede13cbf9da8af32924 Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Wed, 8 Jun 2022 21:09:02 +0530
Subject: [PATCH] ath12k: refactor monitor stats ring processing

Refactor monitor ring reaping logic to collect all the SKBs from the
destination ring and processing the SKB list separately to parse all
the TLVs.

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_mon.c | 119 +++++++++++++++++--------------
 1 file changed, 65 insertions(+), 54 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_mon.c
+++ b/drivers/net/wireless/ath/ath12k/dp_mon.c
@@ -811,7 +811,8 @@ ath12k_dp_mon_rx_parse_status_tlv(struct
 		spin_unlock_bh(&buf_ring->idr_lock);
 
 		if (unlikely(!msdu)) {
-			ath12k_warn(ab, "montior destination with invalid buf_id %d\n",
+			ath12k_dbg(ab, ATH12K_DBG_DATA,
+				   "montior destination with invalid buf_id %d\n",
 				    buf_id);
 			return HAL_RX_MON_STATUS_PPDU_NOT_DONE;
 		}
@@ -1230,7 +1231,11 @@ ath12k_dp_mon_rx_parse_mon_status(struct
 	struct sk_buff *head_msdu, *tail_msdu;
 	enum hal_rx_mon_status hal_status = HAL_RX_MON_STATUS_BUF_DONE;
 
-	ath12k_dp_mon_parse_rx_dest(ab, pmon, skb);
+	hal_status = ath12k_dp_mon_parse_rx_dest(ab, pmon, skb);
+	if (hal_status != HAL_RX_MON_STATUS_PPDU_DONE) {
+		dev_kfree_skb_any(skb);
+		return hal_status;
+	}
 
 	list_for_each_entry_safe(mon_mpdu, tmp, &pmon->dp_rx_mon_mpdu_list, list) {
 		list_del(&mon_mpdu->list);
@@ -2056,18 +2061,15 @@ int ath12k_dp_mon_srng_process(struct at
 	struct dp_srng *mon_dst_ring;
 	struct hal_srng *srng;
 	struct dp_rxdma_ring *buf_ring;
+	struct hal_rx_mon_ppdu_info *ppdu_info = &pmon->mon_ppdu_info;
+	struct sk_buff_head skb_list;
+	struct ath12k_peer *peer = NULL;
 	u64 cookie;
-	u32 ppdu_id, end_reason;
+	u32 hal_status, end_reason, rx_buf_sz, ppdu_id;
 	int num_buffs_reaped = 0, srng_id, buf_id;
-	u8 dest_idx = 0, i;
-	struct hal_rx_mon_ppdu_info *ppdu_info;
-	struct ath12k_peer *peer = NULL;
-	u32 rx_buf_sz;
 	u16 log_type = 0;
 
-	ppdu_info = &pmon->mon_ppdu_info;
-	memset(ppdu_info, 0, sizeof(*ppdu_info));
-	ppdu_info->peer_id = HAL_INVALID_PEERID;
+	__skb_queue_head_init(&skb_list);
 
 	srng_id = ath12k_hw_mac_id_to_srng_id(&ab->hw_params, mac_id);
 
@@ -2088,7 +2090,6 @@ int ath12k_dp_mon_srng_process(struct at
 	}
 
 	srng = &ab->hal.srng_list[mon_dst_ring->ring_id];
-
 	spin_lock_bh(&srng->lock);
 	ath12k_hal_srng_access_begin(ab, srng);
 
@@ -2100,6 +2101,7 @@ int ath12k_dp_mon_srng_process(struct at
 			break;
 
 		cookie = mon_dst_desc->cookie;
+		ppdu_id = mon_dst_desc->ppdu_id;
 		buf_id = u32_get_bits(cookie, DP_RXDMA_BUF_COOKIE_BUF_ID);
 
 		spin_lock_bh(&buf_ring->idr_lock);
@@ -2116,58 +2118,71 @@ int ath12k_dp_mon_srng_process(struct at
 		dma_unmap_single(ab->dev, rxcb->paddr,
 				 skb->len + skb_tailroom(skb),
 				 DMA_FROM_DEVICE);
-
-		pmon->dest_skb_q[dest_idx] = skb;
-		dest_idx++;
-		ppdu_id = mon_dst_desc->ppdu_id;
 		end_reason = u32_get_bits(mon_dst_desc->info0,
-					   HAL_MON_DEST_INFO0_END_REASON);
-		if (end_reason != HAL_MON_END_OF_PPDU) {
+					  HAL_MON_DEST_INFO0_END_REASON);
+		if ((end_reason == HAL_MON_FLUSH_DETECTED) ||
+		    (end_reason == HAL_MON_PPDU_TRUNCATED)) {
 			ath12k_warn(ab, "Monitor dest descriptor end reason %d", end_reason);
-			continue;
+			dev_kfree_skb_any(skb);
+			goto move_next;
 		}
 
-		for (i = 0; i < dest_idx; i++) {
-			skb = pmon->dest_skb_q[i];
+		__skb_queue_tail(&skb_list, skb);
+move_next:
+		ath12k_dp_mon_buf_replenish(ab, buf_ring, 1);
+		ath12k_hal_srng_dst_get_next_entry(ab, srng);
+		num_buffs_reaped++;
+	}
 
-			if (log_type)
-				trace_ath12k_htt_rxdesc(ar, skb->data,
-							log_type, rx_buf_sz);
-
-			if (flag == ATH12K_DP_RX_MONITOR_MODE)
-				ath12k_dp_mon_rx_parse_mon_status(ar, pmon, mac_id,
-								  skb, napi);
-			else
-				ath12k_dp_mon_tx_parse_mon_status(ar, pmon, mac_id,
-								  skb, napi, ppdu_id);
+	ath12k_hal_srng_access_end(ab, srng);
+	spin_unlock_bh(&srng->lock);
 
-			peer = ath12k_peer_find_by_id(ab, ppdu_info->peer_id);
+	if (!num_buffs_reaped)
+		return 0;
 
-			if (!peer || !peer->sta) {
-				ath12k_dbg(ab, ATH12K_DBG_DATA,
-					   "failed to find the peer with peer_id %d\n",
-					   ppdu_info->peer_id);
-				dev_kfree_skb_any(skb);
-				continue;
-			}
+	if (!ppdu_info->ppdu_continuation)
+		memset(ppdu_info, 0, sizeof(*ppdu_info));
 
-			if (ath12k_debugfs_is_pktlog_peer_valid(ar, peer->addr))
-				trace_ath12k_htt_rxdesc(ar, skb->data,
-							log_type, rx_buf_sz);
+	while ((skb = __skb_dequeue(&skb_list))) {
+		if (log_type)
+			trace_ath12k_htt_rxdesc(ar, skb->data,
+						log_type, rx_buf_sz);
+
+		if (flag == ATH12K_DP_RX_MONITOR_MODE)
+			hal_status = ath12k_dp_mon_rx_parse_mon_status(ar, pmon, mac_id,
+								       skb, napi);
+		else
+			hal_status =
+				ath12k_dp_mon_tx_parse_mon_status(ar, pmon, mac_id,
+								  skb, napi, ppdu_id);
 
+		if (hal_status != HAL_RX_MON_STATUS_PPDU_DONE) {
+			ppdu_info->ppdu_continuation = true;
 			dev_kfree_skb_any(skb);
-			pmon->dest_skb_q[i] = NULL;
+			continue;
 		}
 
-		dest_idx = 0;
-move_next:
-		ath12k_dp_mon_buf_replenish(ab, buf_ring, 1);
-		ath12k_hal_srng_src_get_next_entry(ab, srng);
-		num_buffs_reaped++;
-	}
+		rcu_read_lock();
+		spin_lock_bh(&ab->base_lock);
+		peer = ath12k_peer_find_by_id(ab, ppdu_info->peer_id);
+		if (!peer || !peer->sta) {
+			ath12k_dbg(ab, ATH12K_DBG_DATA,
+				   "failed to find the peer with peer_id %d\n",
+				   ppdu_info->peer_id);
+			goto next_skb;
+		}
 
-	ath12k_hal_srng_access_end(ab, srng);
-	spin_unlock_bh(&srng->lock);
+		if (ath12k_debugfs_is_pktlog_peer_valid(ar, peer->addr))
+			trace_ath12k_htt_rxdesc(ar, skb->data,
+						log_type, rx_buf_sz);
+
+next_skb:
+		spin_unlock_bh(&ab->base_lock);
+		rcu_read_unlock();
+		dev_kfree_skb_any(skb);
+		memset(ppdu_info, 0, sizeof(*ppdu_info));
+		ppdu_info->peer_id = HAL_INVALID_PEERID;
+	}
 
 	return num_buffs_reaped;
 }
@@ -2575,7 +2590,7 @@ move_next:
 		peer = ath12k_peer_find_by_id(ab, ppdu_info->peer_id);
 		if (!peer || !peer->sta) {
 			ath12k_dbg(ab, ATH12K_DBG_DATA,
-				   "failed to find the peer with peer_id %d\n",
+				   "failed to find the peer with monitor peer_id %d\n",
 				   ppdu_info->peer_id);
 			goto next_skb;
 		}
--- a/drivers/net/wireless/ath/ath12k/hal_rx.h
+++ b/drivers/net/wireless/ath/ath12k/hal_rx.h
@@ -231,6 +231,7 @@ struct hal_rx_mon_ppdu_info {
 	bool first_msdu_in_mpdu;
 	bool is_ampdu;
 	u8 medium_prot_type;
+	bool ppdu_continuation;
 };
 
 #define HAL_RX_PPDU_START_INFO0_PPDU_ID		GENMASK(15, 0)
